### PR TITLE
Improve notification UI

### LINKED ISSUE #335: Severity for notifications and notification actions
I think it would be good to different levels of importance for the notifications. 

For instance, if i am on the go it is not that important to know that the washer has finished and i don't want to be disturbed because of that. But in case a burglar is triggering the alarm, water is leaking somewhere in the house or it's burning, i always want to notified of course.

As far as i could see there is already a severity parameter in the sendBroadcastNotification call. I guess it is just unused on the android side and has to be implemented. By the way, the upcoming android O release will support so called notification channels, which would fit together very well:
https://developer.android.com/preview/features/notification-channels.html#Importance

Additionally it would be great to have support for notification actions to be able to react to the events instantly. Example: Switch of the siren in case of false alarm. I am not sure how to handle that technically. Maybe you would pass an array of items with label, item and value from from within the rule.

### PR DESCRIPTION
**Pull request by maniac103**
Beautify notification fragment, add summary notifications (API 23+) and implement notification channels (API 26+)

Closes #335 
Also see #345

### CODE REVIEW

**Comment by maniac103** [comment]
This closes #335, even though it also needs openhab-cloud support.

**Comment by mueller-ma** [comment]
Can you bump the target api to 27 in with this PR?

**Comment by maniac103** [comment]
The paged loading of notifications needs #870 in order to work properly.

**Comment by mueller-ma** [comment]
I merged this locally in master and checked it on a Android 24 device. Everything there (except the channels of course) looks good, thought I have tested it with myopenhab, so without https://github.com/openhab/openhab-cloud/pull/175.

**Comment by maniac103** [comment]
Speaking of that PR, I have two additional patches for it: one using the new 'timestamp' field and one using the new 'persistedId' field to change the click handling of notifications (open the notifications page and highlight the respective entry instead of showing an AlertDialog with the notification message). Is it OK to add them here or should I put them in a separate PR?

**Comment by mueller-ma** [comment]
You can put them here, I have to test it again anyway

**Comment by mueller-ma** [comment]
I don't use notifications normally so I tested them with current master before I reviewed this PR. I noticed that when opening the notification list all oh icons are grey, but when scrolling new icons are colored. This seems to be fixed now as all icons are colored. What was the intentional behavior before? Colored or grey icons?

And what about using no default icon or a more [generic icon](https://material.io/tools/icons/?icon=notifications&style=baseline)?

**Comment by maniac103** [comment]
OK, will do so on Tuesday. Need to test it with my local OH cloud instance first.

**Comment by maniac103** [comment]
> This seems to be fixed now as all icons are colored. What was the intentional behavior before? Colored or grey icons?

Huh, I didn't change icon binding. I wonder whether there's some missing mutate() call somewhere, as grey icons sound like those drawables share state with the one in the navigation drawer. In any case, intended behavior is colored icons.
 
> And what about using no default icon or a more generic icon?

Not sure whether having empty space instead of the OH icon would gain much. Do you have some specific alternative icon in mind?

**Comment by maniac103** [comment]
Note to self: probably also should show severity in the notification list if present.

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java [hunk 1]
```diff
@@ -967,16 +968,13 @@ public void onSaveInstanceState(Bundle savedInstanceState) {
...
-        mPendingOpenNotifications = true;
-        openNotificationsPageIfNeeded();
-
-        if (intent.hasExtra(EXTRA_MESSAGE)) {
-            new AlertDialog.Builder(this)
-                    .setTitle(getString(R.string.dlg_notification_title))
-                    .setMessage(intent.getStringExtra(EXTRA_MESSAGE))
-                    .setPositiveButton(getString(android.R.string.ok), null)
-                    .show();
+        mPendingOpenedNotificationId = intent.getStringExtra(EXTRA_NOTIFICATION_ID);
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: I get the following error when using myopenhab:
````
05-23 15:23:07.976 12279-12279/org.openhab.habdroid D/OpenHABMainActivity: Notification was selected
05-23 15:23:07.977 12279-12279/org.openhab.habdroid W/Bundle: Key notificationId expected String but value was a java.lang.Integer.  The default value <null> was returned.
05-23 15:23:08.005 12279-12279/org.openhab.habdroid W/Bundle: Attempt to cast generated internal exception:
    java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
        at android.os.BaseBundle.getString(BaseBundle.java:1121)
        at android.content.Intent.getStringExtra(Intent.java:7139)
        at org.openhab.habdroid.ui.OpenHABMainActivity.onNotificationSelected(OpenHABMainActivity.java:971)
        at org.openhab.habdroid.ui.OpenHABMainActivity.processIntent(OpenHABMainActivity.java:377)
        at org.openhab.habdroid.ui.OpenHABMainActivity.onCreate(OpenHABMainActivity.java:269)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
````
The notification list opens, but nothing is highlighted.

**Comment by mueller-ma** [comment]
I see `05-23 15:22:59.325 12279-12416/org.openhab.habdroid W/EnhancedIntentService: Service took too long to process intent: com.google.android.c2dm.intent.RECEIVE App may get closed.` quite often when receiving notifications.

File: mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java (hunk 1 above)
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/ui/OpenHABMainActivity.java]: Hmm, this probably is the result of the same extra name being used in line 113 and 115 here: https://github.com/openhab/openhab-android/pull/848/files#diff-4f8ecc75b9dbc55f0cf179f386d56722R113 ... I wonder why this didn't happen for me (highlighting works), but different Android versions maybe handle this differently. I'll rename one of the extras.

**Comment by mueller-ma** [comment]
> Is this the case in your testing?

Yes, this was the case. I set my openhab server to "Notifications & Remote Access", however I still see this line in the log. I'm using an emulated Nexus 5 with Android 8.1 and Play Store.

How does it come that I can see the custom icons when using myopenhab? https://github.com/openhab/openhab-cloud/pull/175 isn't merged yet.

Some other things I discovered:
1. The notification channel is always shown as "Default". I know this also needs https://github.com/openhab/openhab-cloud/pull/175, but icons work...
2. The timestamp of the notification summary is the time when the notification is shown, not the google.sent_time. When I expand the notification, the timestamps on the sub-notifications is shown correctly.
3. The icon is not shown in the notification itself like in a messaging app the avatar. Is this intended?

**Comment by mueller-ma** [comment]
> What do you mean exactly - where is it shown?

When I swipe the notification a bit to the right (or left) and click on the gear that appeared, it says "Default" and "1 of 1 notifcation categories from this app".
It's shown correctly in the notification list in the app.

**Comment by maniac103** [comment]
OK, in that case it's expected behavior as outlined above. It will work correctly once my openhab-cloud patch is merged.

**Comment by mueller-ma** [comment]
> Probably should add setWhen() with latest timestamp to summary notification; I'll do that.

Working now :+1: 

> open the notifications page and highlight the respective entry instead of showing an AlertDialog with the notification message

The notification list is opened, but I don't see any highlight.

I have a small commit here to be cherry-picked: https://github.com/openhab/openhab-android/commit/d76cdefede8d545254d97dd521d2c4db883cce6f :cherries:

**Comment by maniac103** [comment]
> The notification list is opened, but I don't see any highlight.

Yes, because the code relies on the 'persistedId' GCM message property provided by my openhab-cloud patch ... maybe we can nag @digitaldan to finally merge it? ;-)

> I have a small commit here to be cherry-picked: d76cdef

Picked, and added another commit deduping the logic on top.

**Pull request merged**